Docs build fixes

mbolivar opened the pull request:
The documentation build works well for non-incremental, CI builds, but has some issues which matter more to the developer's edit/build workflow.

- its content extraction script (extract_content.py) is unnecessarily coarse-grained, considering every single source .rst file instead of just letting the build system copy updated ones after configure time
- extract_content.py doesn't track its source dependencies properly when copying .rst/etc into the build directory, so updated .rst sources don't trigger rebuilds
- its handling of SPHINX_OUTPUT_DIR doesn't look right
- running sphinx-build without `-j` leaves huge opportunities for parallelism on the table
- there's no way for developers to *just* run sphinx-build to update e.g. the HTML if they know what they are doing

This PR addresses these issues. A fresh documentation build is 75% faster on my system. Additionally, careful use of the targets which just re-run sphinx-build allows rebuilding the HTML docs in under 10 seconds.

There's a lot of room for further improvement (in particular, adding logic to only re-run Doxygen and regenerate Kconfig docs when truly necessary), but this felt like a good enough place to start (and makes docs editing feel a lot better from my perspective).

Commit:
.gitignore: only ignore build directories

The "build*" pattern is ignoring any file that begins with
"build". Add a trailing "/" to limit the effect of the pattern
to directories.

This continues to ignore build directories, but allows adding files
whose names begin with build, which will be used later to create a
build.rst file for documenting west's build command.

Signed-off-by: Marti Bolivar <marti@foundries.io>

Commit:
doc: make extract_content.py flake8 clean

Address flake8's complaints.

Signed-off-by: Marti Bolivar <marti@foundries.io>

Commit:
doc: clean up extract_content.py use of environment variables

Minor cleanups:

- Allow the user to run extract.content.py -h with ZEPHYR_BASE unset
  without crashing
- Replace the use of the ZEPHYR_BUILD environment variable with an
  --ignore argument (that can be given multiple times) which specifies
  a path to a directory whose files should *not* be copied, and use
  it to ignore the build directory
- rename, reorder, and tweak get_files() arguments a bit to go from
  source to dest, with extra configuration at the end, getting rid of
  local shadowing of global built-in functions

Signed-off-by: Marti Bolivar <marti@foundries.io>

Commit:
doc: rewrite extract_content.py for better input/output tracking

Separate the phases of creating the list of outputs (and output
directories) from the actual creation of these outputs. This lays the
groundwork for just printing the inputs and outputs in a subsequent
patch, enabling CMake to track dependencies properly.

In the now-separate output generation phase, use shutil.copy2() to
preserve metadata like access and modification times when possible,
rather than re-reading every file and seeing if the contents are
different. This has the same semantics but is clearer and potentially
more efficient, as the initial copies have more accurate relative
timestamps.

Signed-off-by: Marti Bolivar <marti@foundries.io>

Commit:
doc: fix extract_content.py dependency tracking

The logic which copies source documentation files into the build
directory could use some improvements to its dependency management, so
that when a source file changes, extract_content.py gets re-run.

Make these changes as follows:

- Add an --outputs flag to extract_content.py, so that the
  sources it depends on and generates can be saved into a file and
  thus made known to the build system

- Change the way the sources and destination are specified in the
  extract_content.py command line so that the entire job can be done
  in a single command, rather than multiple (to avoid having to
  collate multiple --outputs files in CMake)

- Extract the content at configure time with execute_process(),
  tracking all inputs and outputs within the build system itself. Use
  this information to make sure that each individual output depends on
  just its exact input file, ensuring updated inputs produce updated
  outputs without having to call extract_content.py again.

- Ensure that the "content" build system target depends on all the
  outputs, transitively triggering a rebuild any time an input
  file (e.g. .rst documentation file or included image/source file)
  changes.

Signed-off-by: Marti Bolivar <marti@foundries.io>

Commit:
doc: add comment describing SPHINX_OUTPUT_DIR behavior

Explain the asymmetry with the situation when it is unset.

Signed-off-by: Marti Bolivar <marti@foundries.io>

Commit:
doc: add sphinx-html and sphinx-latex convenience targets

Most of time (i.e. when just writing to .rst files or editing images,
rather than editing Doxygen or Kconfig), a full rebuild of the docs
pipeline is not necessary.

This is unfortunate, because it takes over a minute to re-run doxygen
etc. (on my system), whereas an incremental sphinx-build run should
only take a few seconds.

As a band-aid to help developers who have already generated a docs
build and know the consequences of their actions, add "sphinx-html"
and "sphinx-latex" targets. These just re-extract content files into
the build directory and run sphinx-build. This makes for much faster
documentation edit/compile/read loops.

Add some extra COMMENTs while we're here, to make it more obvious
what's going on when the docs build system crunches on something.

A "proper" solution for changing "ninja html" to have similar
performance would need, at least, to:

- only re-run doxygen if any source files have changed
- only re-generate Kconfig if any of those files have changed

That's all left to future work, as this is good enough to make
rebuilding the docs bearable for me.

Signed-off-by: Marti Bolivar <marti@foundries.io>

SebastianBoe commented on .gitignore:
```diff
@@ -7,7 +7,7 @@
 *.swp
 *.swo
 *~
-build*
```
Since you are touching this, I have observed that this matches things it shouldn't.

E.g. I had some confusing behaviour when I worked in the test tests/drivers/build_all/.

Is it possible to rename it such that it doesn't match things we want to version control?

build_docs_*/ for instance

mbolivar replied:
What if we just ignore `build/` ?

carlescufi commented on doc/CMakeLists.txt:
```diff
@@ -41,9 +41,9 @@ if(NOT DEFINED SPHINX_OUTPUT_DIR)
-  set(SPHINX_OUTPUT_DIR_HTML ${SPHINX_OUTPUT_DIR})
-  set(SPHINX_OUTPUT_DIR_LATEX ${SPHINX_OUTPUT_DIR})
-  set(SPHINX_OUTPUT_DIR_PDF ${SPHINX_OUTPUT_DIR})
+  set(SPHINX_OUTPUT_DIR_HTML ${SPHINX_OUTPUT_DIR}/html)
```
This will break our downstream builds, since we need to be able to specify exactly where the HTML is placed because we use intersphinx after to combine with our own doc.

carlescufi replied:
I propose we instead let users specify SPHINX_OUTPUT_DIR_HTML directly

mbolivar replied:
I was wondering about this. I'll just replace this patch with a comment describing what is going on. Thanks for the explanation.

carlescufi commented on doc/CMakeLists.txt:
```diff
@@ -68,9 +68,76 @@ set(DOC_LOG ${CMAKE_CURRENT_BINARY_DIR}/doc.log)
+  list(APPEND EXTRACT_CONTENT_OUTPUTS ${DST})
+endwhile()
+
+add_custom_target(content DEPENDS ${EXTRACT_CONTENT_OUTPUTS})
```
What if we add a new .rst file to the `doc/` folder or elsewhere? do we need then a clean build for it to be included?

mbolivar replied:
You just need to re-run cmake. The execute_process() will then run again, and the dependencies will be updated. This is basically the same thing that happens in source code that uses GLOB to find its C files.

carlescufi replied:
This should be documented in `doc/README` then I think, since CMake won't be run automatically.

carlescufi reviewed (changes requested):
Thanks for all the great improvements @mbolivar!
There's a couple of things that need clarifying or fixing but otherwise this is indeed much better than the previous iteration.

carlescufi commented on doc/CMakeLists.txt:
```diff
@@ -31,7 +31,7 @@ if(${LATEXMK} STREQUAL LATEXMK-NOTFOUND)
 
 if(NOT DEFINED SPHINXOPTS)
-  set(SPHINXOPTS -q)
+  set(SPHINXOPTS -j auto -q)
```
I tried this but there were some issues with the generated content. Are you sure this generates the very same output?

dbkinder replied:
I pulled @mbolivar's patch and generated content locally both with and without the ``-j auto`` on the sphinxbuild command.  There are differences in 10 generated html files, but as I look at the html diffs, the differences are curious but not consequential -- the generated output looks the same (or very close) in the web browser. So, the earlier issues with ``-j auto`` identified in other PRs/Issues didn't show up.  This is encouraging.

mbolivar replied:
Same story here. The build produces slightly different HTML but I don't observe differences in the browser. I can drop this part of the series if you prefer.

mbolivar replied:
Here's the sequential --> parallel diff.

I took a look at the files in a browser (including a board where the images were different, and the bluetooth page, which had different sidebar HTML) and didn't see any behavioral or visual differences.

mbolivar replied:
> For completion @mbolivar, I would be grateful if you could run this on Windows and verify that you also get the same output before setting it to default

I have done this and looked at the results in the same files I checked on Linux (bluetooth.html, the sensor API, Nucleo F207ZG) and they seem fine on Windows as well.

nashif commented:
I tried that with a theme (a git repo) checked out into doc/themes, got some permission issues when I was trying to rebuild the docs. I meanwhile remove the repo and do not have the build output... Not sure what was going on.

Commit:
doc: cmake must be re-run if adding/removing files

This is a consequence of the move to tracking each documentation
source file individually.

Signed-off-by: Marti Bolivar <marti@foundries.io>

carlescufi reviewed (approved):
My comments have been addressed in full.

mbolivar commented:
Weird, shippable's build of the docs is complaining because of https://app.shippable.com/github/zephyrproject-rtos/zephyr/runs/24296/5/tests, which reads:

```
/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/doc/_build/rst/doc/subsystems/sensor.rst: WARNING: Duplicate declaration, also defined in 'api/io_interfaces'.
Declaration is 'sensor_value'.
```

But that appears to be a known issue in .known-issues/doc/duplicate.conf.

Going to see if I can change the regexp in case the "also defined in" is what's the trouble.

mbolivar commented:
> Going to see if I can change the regexp in case the "also defined in" is what's the trouble.

I can't figure this out and I don't have time to debug further. Dropping the -j auto part for this PR. Oh well, at least we know it's usable.

carlescufi commented:
@dbkinder are you happy with this as it is now?

dbkinder reviewed (approved):
no -j then for now.  OK.

marc-hb commented:
Thanks @gmarull for your patience making some adjustments, preserving some features and thoroughly explaining and documenting everything. We had a look at 2 years-old PRs, I think this proves the latter is useful :-)

Not going to hit the Approve button because I barely looked at the code but overall LGTM.

2 nits:
- Maybe there should be a bit more verbose comments in the code and especially with respect to performance. I mean comments copied straight from this PR discussion. Because the doc build is unfortunately a bit complicated and its performance aspects even less intuitive, I think the current discussions demonstrate that. So I suggest reviewing every explanation you gave in this PR and gauge whether it's source-worthy (if not in a comment yet).
- rsync/dirsync function, see below.


> The modification time script becomes mostly redundant once input dependency tracking is introduced. With that feature in place, Doxygen is only run when any of the input file is changed.

Finally got it, thanks.

> There is an edge case that the script could solve: what if I edit .c/.h, but I don't change any docstring? In theory the generated XML files should be the same, so I could skip updating the XML files given to Sphinx. We can add this case to the extension, I wonder though if it is worth it.

I don't think it is such an edge case, I think it's easy to imagine someone multi tasking between `.rst` documentation work and C coding that does not involve doxygen changes.

We should also hope that the the "everything depends on everything" breathe bug might get fixed some day. Then smart doxygen output with correct timestamps should make a huge difference because breathe is so time consuming.

Also keep in mind developers' patience is extremely limited with respect to documentation, much more limited than with code. So every little helps!

Speaking of developer (im)patience and incremental build performance, here's one simple CI suggestion: after a succesful doc build in CI, perform a single change in a single .rst file and build again. Time the second, incremental build, fail CI if that takes more than 10 or 15 seconds. Same again with a single change to a single C file (with a higher threshold).
